### PR TITLE
Add Learning Rate Scheduler C API

### PR DESCRIPTION
**Pull request by baijumeswani**
This pull request adds an API for registering a scheduler with the training session.

```cc
// c api
OrtTrainingSession* session;
g_ort_training_api->CreateTrainingSession(env, session_options, checkpoint_state,
                                          training_model.c_str(), eval_model.c_str(),
                                          optimizer_model.c_str(), &session));

// Set the initial learning rate.
g_ort_training_api->SetLearningRate(session, learning_rate);

// Register the linear leaning rate scheduler.
g_ort_training_api->RegisterLinearLRScheduler(session, warm_up_step_count, total_step_count);

// Take a scheduler step.
g_ort_training_api->SchedulerStep(session);
```

### CODE REVIEW

**Review by pengwa — DISMISSED** [review]
LGTM

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h [hunk 1]
```diff
@@ -144,6 +144,20 @@ struct OrtTrainingApi {
                   size_t inputs_len, _In_reads_(inputs_len) const OrtValue* const* inputs,
                   size_t outputs_len, _Inout_updates_all_(outputs_len) OrtValue** outputs);
 
+  /** \brief Set the learning rate for this training session.
```
Comment (pengwa) [inline orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:147]: maybe use "Sets" since other comments do like this..

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h [hunk 2]
```diff
@@ -161,6 +175,36 @@ struct OrtTrainingApi {
...
+  * from 0 to the initial learning rate set on the training session.
+  *
+  * \param[in] sess The training session that should use the linear learning rate scheduler.
+  * \param[in] warmup_step_count The number of steps in the warm up phase.
+  * \param[in] total_step_count The total number of training steps.
+  *
+  * \snippet{doc} snippets.dox OrtStatus Return Value
+  *
+  */
+  ORT_API2_STATUS(RegisterLinearLRScheduler, _Inout_ OrtTrainingSession* sess, _In_ int64_t warmup_step_count,
```
Comment (pengwa) [inline orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:192]: there will be quite a few different scheduler, I see from user repo. do you expect we add RegisterXXX for every one? 

Or mayb we can pass the scheduler type and required parameters with a single register interface?

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h [hunk 3]
```diff
@@ -144,6 +144,20 @@ struct OrtTrainingApi {
...
+  * session by default starts with a learning rate of 0.001. It can be overwritten by invoking
+  * this function with the desired learning rate.
+  *
+  * \param[in] sess The training session on which the learning rate needs to be set.
+  * \param[in] learning_rate Desired learning rate to set.
+  *
+  * \snippet{doc} snippets.dox OrtStatus Return Value
+  *
+  */
+  ORT_API2_STATUS(SetLearningRate, _Inout_ OrtTrainingSession* sess, _In_ float learning_rate);
```
Comment (pengwa) [inline orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:159]: I think it's better we call it InitialLearningRate?

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h [hunk 4]
```diff
@@ -144,6 +144,20 @@ struct OrtTrainingApi {
                   size_t inputs_len, _In_reads_(inputs_len) const OrtValue* const* inputs,
                   size_t outputs_len, _Inout_updates_all_(outputs_len) OrtValue** outputs);
 
+  /** \brief Set the learning rate for this training session.
+  *
+  * This function allows users to set the learning rate for the training session. The training
+  * session by default starts with a learning rate of 0.001. It can be overwritten by invoking
```
Comment (pengwa) [inline orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:150]: nit: listing the 0.001 value could be easily lost of track once the default learning rate get changed (in another file).

File: orttraining/orttraining/training_api/training_session.cc [hunk 5]
```diff
@@ -21,6 +21,14 @@ TrainingSession::TrainingSession(const Environment& session_env,
                            session_options, session_env, providers)
                      : std::unique_ptr<Optimizer>()} {}
 
+Status TrainingSession::RegisterScheduler(const std::function<
+                                          std::unique_ptr<LRSchedulerBase>(std::shared_ptr<Optimizer>)>& get_scheduler) {
```
Comment (pengwa) [inline orttraining/orttraining/training_api/training_session.cc:25]: I belive you already notice those cpplint issue.

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h (hunk 2 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:192]: The problem with a single register function is that it will lead to bloated parameters to the function. We might have input parameters that are not relevant to that particular scheduler in the function signature. Even if we put it in a struct, the user might find it confusing to pick the correct arguments for their specific use case.

One solution to this problem is to expose structs specific for each scheduler. For example, to use the LinearLRScheduler, users must pass in a `void*` to a struct of type `LinearLRSchedulerParameters` along with the enum that we expose for `LinearLRScheduler`. We can type cast it to the correct type inside the implementation and accordingly construct the correct C++ object.

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h (hunk 4 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:150]: How else shall we document the default learning rate? It might not be convenient for users to read the `Optimizer` cpp implementation to find the default learning rate.

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h (hunk 3 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:159]: Do we envision users building their own lr scheduler and periodically invoking `SetLearningRate`? If yes, we should probably call it `SetLearningRate`. If we think this function should only be called to set the initial learning rate, we could then call it `InitialLearningRate`. What do you think?

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h (hunk 2 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h]: I made the change so we now take input a `void*` to the learning rate scheduler parameters and an enum. Let me know what you think.

File: orttraining/orttraining/training_api/onnxruntime_training_c_api.cc [hunk 6]
```diff
@@ -210,16 +216,34 @@ ORT_API_STATUS_IMPL(OrtTrainingApis::OptimizerStep, _Inout_ OrtTrainingSession*
...
-  }));
+  switch (lr_scheduler_type) {
+    case OrtLRSchedulerType::LinearLRScheduler: {
+      auto parameters = reinterpret_cast<OrtLinearLRSchedulerParameters*>(lr_scheduler_parameters);
+      ORT_API_RETURN_IF_STATUS_NOT_OK(session->RegisterScheduler([&parameters](auto optimizer) {
+        return std::make_unique<onnxruntime::training::api::LinearLRScheduler>(
+            optimizer, parameters->warmup_step_count, parameters->total_step_count);
+      }));
+
+      if (initial_lr) {
```
Comment (pengwa) [inline orttraining/orttraining/training_api/onnxruntime_training_c_api.cc:235]: For our API, if a nullptr initial_lr is given, should give an explictily error message?

File: orttraining/orttraining/training_api/onnxruntime_training_c_api.cc [hunk 7]
```diff
@@ -210,16 +216,34 @@ ORT_API_STATUS_IMPL(OrtTrainingApis::OptimizerStep, _Inout_ OrtTrainingSession*
...
 
+  OrtStatus* status = nullptr;
+
   auto session = reinterpret_cast<onnxruntime::training::api::TrainingSession*>(sess);
-  ORT_API_RETURN_IF_STATUS_NOT_OK(session->RegisterScheduler([warmup_step_count, total_step_count](auto optimizer) {
-    return std::make_unique<onnxruntime::training::api::LinearLRScheduler>(optimizer, warmup_step_count, total_step_count);
-  }));
+  switch (lr_scheduler_type) {
+    case OrtLRSchedulerType::LinearLRScheduler: {
+      auto parameters = reinterpret_cast<OrtLinearLRSchedulerParameters*>(lr_scheduler_parameters);
```
Comment (pengwa) [inline orttraining/orttraining/training_api/onnxruntime_training_c_api.cc:229]: I think you need also check whether parameters is nullptr or now (considering user probabbly give a cosine_schduler_parameters).

Comment (pengwa) [reply orttraining/orttraining/training_api/onnxruntime_training_c_api.cc:229]: Oh Okay,  I thought we are using dynamic_cast. I feel it is dangerous here we probably cast a cosine parameter pointer  to  linear parameter pointer without check?

File: orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h (hunk 3 above)
Comment (pengwa) [reply orttraining/orttraining/training_api/include/onnxruntime_training_c_api.h:168]: I see, thanks for the explanation! I think this is fine, let's keep the name.

File: orttraining/orttraining/training_api/onnxruntime_training_c_api.cc (hunk 6 above)
Comment (pengwa) [reply orttraining/orttraining/training_api/onnxruntime_training_c_api.cc:235]: Oh I see, it is an optional param.

File: orttraining/orttraining/training_api/onnxruntime_training_c_api.cc [hunk 8]
```diff
@@ -184,6 +190,17 @@ ORT_API_STATUS_IMPL(OrtTrainingApis::EvalStep, _In_ const OrtTrainingSession* se
   API_IMPL_END
 }
 
+ORT_API_STATUS_IMPL(OrtTrainingApis::SetLearningRate, _Inout_ OrtTrainingSession* sess,
+                    _In_ float learning_rate) {
+  API_IMPL_BEGIN
+
+  auto session = reinterpret_cast<onnxruntime::training::api::TrainingSession*>(sess);
+  ORT_API_RETURN_IF_STATUS_NOT_OK(SetLearningRateImpl(session, learning_rate));
```
Comment (pengwa) [inline orttraining/orttraining/training_api/onnxruntime_training_c_api.cc:198]: Sorry, get back here again. :)

I think this is still doing the same thing in term of  setting learning rate as RegisterLRScheduler did when initial_lr is given. 
They are both operating on internal initial_lr. So shall we change the parameter name "learning_rate" to "initial_lr" for alignment.

Comment (pengwa) [reply orttraining/orttraining/training_api/onnxruntime_training_c_api.cc:198]: OKay, I know the gap now. The private SetLearningRate in orttraining/orttraining/training_api/include/optimizer.h is internally used by lr_shcudler during Step(), the parameter it takes is the lr after calculation based on initial_lr and some factors. 

Here, when we register LR with an initial_lr, we want to set initial_lr. So we need a public  SetInitialLearningRate(float initial_lr) interface for Optimizer. 

I would suggest for OrtTrainingApis::SetLearningRate, we also set the initial lr, I believe when user want to set lr, they usually means the initial lr.

File: orttraining/orttraining/training_api/include/training_session.h [hunk 9]
```diff
@@ -46,14 +50,19 @@ class TrainingSession {
...
+
   Status CreateCheckpointState(CheckpointState& chkpt_state, bool save_optimizer_state) const;
 
  private:
   ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(TrainingSession);
 
   const std::unordered_map<std::string, std::shared_ptr<Parameter>> named_parameters_;
   std::unique_ptr<Module> module_;
-  std::unique_ptr<Optimizer> optimizer_;
+  std::shared_ptr<Optimizer> optimizer_;
```
Comment (askhade) [inline orttraining/orttraining/training_api/include/training_session.h:64]: why shared?

File: orttraining/orttraining/training_api/onnxruntime_training_c_api.cc (hunk 7 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/onnxruntime_training_c_api.cc:229]: dynamic_cast might be useful for up casts and down casts. In this case, we want to cast the pointer back to the original type based on the enum. We rely on the user to provide the correct enum.

File: orttraining/orttraining/training_api/include/training_session.h (hunk 9 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/include/training_session.h:64]: the optimizer is now shared between the lr scheduler and the training session. The `lr_scheduler.Step()` will set the new learning rate on the optimizer it co-owns.

File: orttraining/orttraining/training_api/onnxruntime_training_c_api.cc (hunk 8 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/onnxruntime_training_c_api.cc:198]: Seems like the initial learning rate only matters for pre-defined lr schedulers. What we can do is allow users to set the initial learning rate when registering the learning rate scheduler using the `RegisterLRScheduler` api.
And if they want to set the learning rate (not the initial one), they can use the `SetLearningRate` api for scenarios where they have custom lr schedulers.

**Review by pengwa — APPROVED** [review]

File: orttraining/orttraining/training_api/onnxruntime_training_c_api.cc (hunk 8 above)
Comment (pengwa) [reply orttraining/orttraining/training_api/onnxruntime_training_c_api.cc]: Okay, that means, for user defined lr scheduler, they can use SetLearningRate, does it imply RegisterLRScheduler is not needed for this kind of usage?

File: orttraining/orttraining/training_api/training_session.cc [hunk 10]
```diff
@@ -58,6 +71,16 @@ Status TrainingSession::CreateCheckpointState(CheckpointState& chkpt_state, bool
...
 }
 
+Status TrainingSession::SetLearningRate(float learning_rate) noexcept {
+  ORT_RETURN_IF_ERROR(optimizer_->SetLearningRate(learning_rate));
+
+  return Status::OK();
+}
+
+Status TrainingSession::SchedulerStep() noexcept {
+  return scheduler_->Step();
```
Comment (pengwa) [inline orttraining/orttraining/training_api/training_session.cc:81]: if RegisterScheduler is not required when user want to use customized lr scheduler, then scheduler_ will be null, right?

Comment (baijumeswani) [reply orttraining/orttraining/training_api/training_session.cc:81]: That is right. I think it makes sense to check for nullptr here. Thanks.

File: orttraining/orttraining/training_api/onnxruntime_training_c_api.cc (hunk 8 above)
Comment (baijumeswani) [reply orttraining/orttraining/training_api/onnxruntime_training_c_api.cc]: Yes, in case users are using their own LR scheduler, they no longer need to register the learning rate scheduler. They must call `SetLearningRate` based on the custom scheduler they are using.

Comment (pengwa) [reply orttraining/orttraining/training_api/onnxruntime_training_c_api.cc]: Please kindly help add some pseudo-code for how the code looks like when they wanted to use customized lr scheduler.

**Review by pengwa — APPROVED** [review]

**Comment by baijumeswani** [comment]
Thanks for the review.
